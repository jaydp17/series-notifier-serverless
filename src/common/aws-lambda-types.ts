/**
 * Contains types that are related to AWS Lambda
 */

export type LambdaEvent = {
  httpMethod: string;
  body: string;
  queryStringParameters: {} | null;
};

/**
 * Callback type that should be used when the lambda is gonna be invoked with an HTTP req
 */
export type LambdaHttpCallback = (error: Error | null | undefined, response?: { statusCode: number, body?: string }) => void;

/**
 * Non-HTTP lambda callback type
 */
export type LambdaCallback = (error: Error | null | undefined, response?: {}) => void;
