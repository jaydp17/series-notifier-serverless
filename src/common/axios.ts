/**
 * Gives a common axios instance across the entire app
 */

import axios, { AxiosRequestConfig, AxiosResponse } from 'axios';
import Logger from '../common/logger';

const timeLogger = Logger('TimeLogger');

function requestInterceptor(config: AxiosRequestConfig) {
  return { ...config, startTime: Date.now() };
}

function responseInterceptor(response: AxiosResponse) {
  const { startTime } = response.config as any; // tslint:disable-line:no-any
  const endTime = Date.now();
  timeLogger.log(`[TimeLogger]:: ${response.config.url} -> ${endTime - startTime}ms`);
  return response;
}

axios.interceptors.request.use(requestInterceptor);
axios.interceptors.response.use(responseInterceptor);

export default axios;
