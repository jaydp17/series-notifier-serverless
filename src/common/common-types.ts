/**
 * This file contains types that are consumed by more than 1 function
 */

export type Callback = (error: Error | null | undefined, data?: any) => void; // tslint:disable-line:no-any
