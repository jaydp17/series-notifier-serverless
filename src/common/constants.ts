/**
 * Contains constants used accross the app
 */
export const GRAPH_API_URL = 'https://graph.facebook.com/v2.6';
