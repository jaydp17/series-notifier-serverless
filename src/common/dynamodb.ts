/**
 * Creates a dynamodb instance
 */

import { DynamoDB } from 'aws-sdk';
import { env } from './environment';

let dynamoDbOptions = {};
if (env === 'development' && process.env.IS_LOCAL) {
  dynamoDbOptions = { endpoint: 'http://localhost:8000' }; //tslint:disable-line
}
export const dynamodb = new DynamoDB(dynamoDbOptions);
export const dynamoDocClient = new DynamoDB.DocumentClient(dynamoDbOptions);

export default {
  get: params => dynamoDocClient.get(params).promise(),
  put: params => dynamoDocClient.put(params).promise(),
  query: params => dynamoDocClient.query(params).promise(),
  delete: params => dynamoDocClient.delete(params).promise(),
};
