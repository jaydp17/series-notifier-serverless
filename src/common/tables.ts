/**
 * Stores the table names & specs
 */

import { DynamoDB } from 'aws-sdk';

export const names = {
  users: 'Users',
  subscriptions: 'Subscriptions',
};

const usersTable = {
  TableName: 'Users',
  KeySchema: [
    { AttributeName: 'socialId', KeyType: 'HASH' },  //Partition key
  ],
  AttributeDefinitions: [
    { AttributeName: 'socialId', AttributeType: 'S' },
  ],
  ProvisionedThroughput: {
    ReadCapacityUnits: 1,
    WriteCapacityUnits: 1,
  },
};

export const indexes = {
  subscriptions: {
    socialId: 'socialId-index',
  },
};

const subscriptionsTable: DynamoDB.CreateTableInput = {
  TableName: 'Subscriptions',
  KeySchema: [
    { AttributeName: 'imdbId', KeyType: 'HASH' },  //Partition key
    { AttributeName: 'socialId', KeyType: 'RANGE' },  //Sort key
  ],
  AttributeDefinitions: [
    { AttributeName: 'imdbId', AttributeType: 'S' },
    { AttributeName: 'socialId', AttributeType: 'S' },
  ],
  GlobalSecondaryIndexes: [{
    IndexName: indexes.subscriptions.socialId,
    KeySchema: [
      { AttributeName: 'socialId', KeyType: 'HASH' },  //Partition key
      { AttributeName: 'imdbId', KeyType: 'RANGE' },  //Sort key
    ],
    Projection: {
      ProjectionType: 'KEYS_ONLY',
    },
    ProvisionedThroughput: {
      ReadCapacityUnits: 1,
      WriteCapacityUnits: 1,
    },
  }],
  ProvisionedThroughput: {
    ReadCapacityUnits: 1,
    WriteCapacityUnits: 1,
  },
};

export const specs = {
  users: usersTable,
  subscriptions: subscriptionsTable,
};

export default { names, specs, indexes };
