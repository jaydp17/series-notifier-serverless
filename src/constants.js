export const GRAPH_API_URL = 'https://graph.facebook.com/v2.6';
