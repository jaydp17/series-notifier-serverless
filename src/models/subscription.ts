/**
 * Subscription model
 */

import { DynamoDB } from 'aws-sdk';
import dynamodb from '../common/dynamodb';
import tables from '../common/tables';
import { isValidSocialId } from './user';

import { prettyPrint } from '../common/common-utils';

const TableName = tables.names.subscriptions; // tslint:disable-line:variable-name

export type SubscriptionRecord = {
  imdbId: string;
  socialId: string;
};

/**
 * Adds a mapping of imdbId and socialId of a user,
 * to mark the subscription of a given TvShow to a user
 */
export function createSubscription(imdbId: string, socialId: string) {
  if (!isValidSocialId(socialId)) {
    throw new Error(`Invalid socialId: ${socialId}`);
  }
  const params: DynamoDB.DocumentClient.PutItemInput = {
    TableName,
    Item: {
      imdbId,
      socialId,
    },
  };
  return dynamodb.put(params);
}

/**
 * Get all the shows subscribed by a user
 */
export async function getSubscribedShows(socialId: string): Promise<SubscriptionRecord[]> {
  if (!isValidSocialId(socialId)) {
    throw new Error(`Invalid socialId: ${socialId}`);
  }
  const params: DynamoDB.DocumentClient.QueryInput = {
    TableName,
    IndexName: tables.indexes.subscriptions.socialId,
    KeyConditionExpression: 'socialId = :socialId',
    ExpressionAttributeValues: { ':socialId': socialId },
  };
  const result = await dynamodb.query(params);
  return <SubscriptionRecord[]>result.Items;
}

/**
 * Get all the users who have subscribed to a given show
 */
export async function getUsersWhoSubscribed(imdbId: string): Promise<SubscriptionRecord[]> {
  const params: DynamoDB.DocumentClient.QueryInput = {
    TableName,
    KeyConditionExpression: 'imdbId = :imdbId',
    ExpressionAttributeValues: { ':imdbId': imdbId },
  };
  const result = await dynamodb.query(params);
  return <SubscriptionRecord[]>result.Items;
}

/**
 * Removes the subscription of a user from a tvShow
 */
export function deleteSubscription(imdbId: string, socialId: string) {
  const params: DynamoDB.DocumentClient.DeleteItemInput = {
    TableName,
    Key: { imdbId, socialId },
  };
  return dynamodb.delete(params);
}
