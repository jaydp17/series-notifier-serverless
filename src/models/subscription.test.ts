/**
 * Tests for Subscription Model
 */

jest.mock('../common/dynamodb.ts');

// mocks
import dynamodb from '../common/dynamodb';

// imports
import { platformNames } from '../common/constants';
import tables from '../common/tables';
import * as SubscriptionModel from './subscription';

describe('Subscription Model', () => {
  beforeEach(() => {
    jest.resetAllMocks();
  });

  describe('create subscription', () => {
    it('throws error on invalid socialId', () => {
      const socialId = 'asdf';
      const imdbId = 't23454';
      expect(() => {
        SubscriptionModel.createSubscription(imdbId, socialId);
      }).toThrowError(/Invalid socialId/);
    });

    it('creates a new subscription', async () => {
      // prepare
      const imdbId = 't23456';
      const socialId = `${platformNames.FBMessenger}::123`;

      // execute
      await SubscriptionModel.createSubscription(imdbId, socialId);

      // test
      expect(dynamodb.put).toHaveBeenCalledTimes(1);
      expect((<jest.Mock<{}>>dynamodb.put).mock.calls[0]).toMatchSnapshot();
    });
  });

  describe('get all Subscribed shows', () => {
    it('throws error on invalid socialId', async () => {
      // TODO: refactor this when jest 20.0.0+ comes out
      // use expect().rejects
      const socialId = 'asdf';
      try {
        await SubscriptionModel.getSubscribedShows(socialId);
        return Promise.reject(new Error('should have failed'));
      } catch (err) {
        expect(err.message).toContain('Invalid socialId');
      }
    });

    it('gets subscribed shows', async () => {
      // prepare
      const socialId = `${platformNames.FBMessenger}::123`;
      const expectedItems = [{ hello: 'world' }];
      (<jest.Mock<{}>>dynamodb.query).mockReturnValueOnce({ Items: expectedItems });

      // execute
      const items = await SubscriptionModel.getSubscribedShows(socialId);

      // test
      expect(items).toEqual(expectedItems);
      expect(dynamodb.query).toHaveBeenCalledTimes(1);
      expect((<jest.Mock<{}>>dynamodb.query).mock.calls[0]).toMatchSnapshot();
    });
  });

  it('gets users who subscribed', async () => {
    // prepare
    const imdbId = 't123123';
    const expectedItems = [{ hello: 'world' }];
    (<jest.Mock<{}>>dynamodb.query).mockReturnValueOnce({ Items: expectedItems });

    // execute
    const items = await SubscriptionModel.getUsersWhoSubscribed(imdbId);

    // test
    expect(items).toEqual(expectedItems);
    expect(dynamodb.query).toHaveBeenCalledTimes(1);
    expect((<jest.Mock<{}>>dynamodb.query).mock.calls[0]).toMatchSnapshot();
  });

  it('deletes a subscription', async () => {
    // prepare
    const imdbId = 't123123';
    const socialId = `${platformNames.FBMessenger}::123`;

    // execute
    const items = await SubscriptionModel.deleteSubscription(imdbId, socialId);

    // test
    expect(dynamodb.delete).toHaveBeenCalledTimes(1);
    expect((<jest.Mock<{}>>dynamodb.delete).mock.calls[0]).toMatchSnapshot();
  });
});
