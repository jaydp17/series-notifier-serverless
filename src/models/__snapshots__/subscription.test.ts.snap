// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Subscription Model create subscription creates a new subscription 1`] = `
Array [
  Object {
    "Item": Object {
      "imdbId": "t23456",
      "socialId": "FBMessenger::123",
    },
    "TableName": "Subscriptions",
  },
]
`;

exports[`Subscription Model deletes a subscription 1`] = `
Array [
  Object {
    "Key": Object {
      "imdbId": "t123123",
      "socialId": "FBMessenger::123",
    },
    "TableName": "Subscriptions",
  },
]
`;

exports[`Subscription Model get all Subscribed shows gets subscribed shows 1`] = `
Array [
  Object {
    "ExpressionAttributeValues": Object {
      ":socialId": "FBMessenger::123",
    },
    "IndexName": "socialId-index",
    "KeyConditionExpression": "socialId = :socialId",
    "TableName": "Subscriptions",
  },
]
`;

exports[`Subscription Model gets users who subscribed 1`] = `
Array [
  Object {
    "ExpressionAttributeValues": Object {
      ":imdbId": "t123123",
    },
    "KeyConditionExpression": "imdbId = :imdbId",
    "TableName": "Subscriptions",
  },
]
`;
